### PR TITLE
feat: add flag push

### PR DESCRIPTION
**Pull request by zhukaihan**
Add flag push to JVM SDK. This enables streaming flag changes to clients, instead of polling every x interval.

### CODE REVIEW

File: src/main/kotlin/deployment/DeploymentRunner.kt [hunk 1]
```diff
@@ -2,22 +2,19 @@
 
 package com.amplitude.experiment.deployment
 
-import com.amplitude.experiment.ExperimentalApi
-import com.amplitude.experiment.LocalEvaluationConfig
-import com.amplitude.experiment.LocalEvaluationMetrics
+import com.amplitude.experiment.*
```
Comment (zhukaihan) [inline src/main/kotlin/deployment/DeploymentRunner.kt:5]: Was there a discouragement of using * imports?

File: src/main/kotlin/flag/FlagConfigStreamApi.kt [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    deploymentKey: String,
+    serverUrl: HttpUrl,
+    httpClient: OkHttpClient = OkHttpClient(),
+    val connectionTimeoutMillis: Long = CONNECTION_TIMEOUT_MILLIS_DEFAULT,
+    keepaliveTimeoutMillis: Long = KEEP_ALIVE_TIMEOUT_MILLIS_DEFAULT,
+    reconnIntervalMillis: Long = RECONN_INTERVAL_MILLIS_DEFAULT
+) {
+    var onInitUpdate: ((List<EvaluationFlag>) -> Unit)? = null
+    var onUpdate: ((List<EvaluationFlag>) -> Unit)? = null
+    var onError: ((Exception?) -> Unit)? = null
```
Comment (zhukaihan) [inline src/main/kotlin/flag/FlagConfigStreamApi.kt:39]: These callbacks can be directly set. Should I add setters or direct access is fine?

File: src/main/kotlin/deployment/DeploymentRunner.kt [hunk 3]
```diff
@@ -39,21 +38,29 @@ internal class DeploymentRunner(
         null
     }
     private val cohortPollingInterval: Long = getCohortPollingInterval()
+    // Fallback in this order: proxy, stream, poll.
+    private val amplitudeFlagConfigPoller = FlagConfigFallbackRetryWrapper(
+        FlagConfigPoller(flagConfigApi, flagConfigStorage, cohortLoader, cohortStorage, config, metrics),
+        null, config.flagConfigPollerIntervalMillis, 1000
+    )
```
Comment (bgiori) [inline src/main/kotlin/deployment/DeploymentRunner.kt:45]: NIT: The standard newline formatting for kotlin is different from java (more like javascript) where each argument gets its own newline. E.g.

```
private val amplitudeFlagConfigPoller = FlagConfigFallbackRetryWrapper(
        FlagConfigPoller(flagConfigApi, flagConfigStorage, cohortLoader, cohortStorage, config, metrics),
        null, 
        config.flagConfigPollerIntervalMillis, 
        1000
    )
```

File: src/main/kotlin/flag/FlagConfigUpdater.kt [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.withLock
+import kotlin.math.max
+import kotlin.math.min
+
+internal interface FlagConfigUpdater {
+    // Start the updater. There can be multiple calls.
+    // If start fails, it should throw exception. The caller should handle error.
+    // If some other error happened while updating (already started successfully), it should call onError.
+    fun start(onError: (() -> Unit)? = null)
```
Comment (bgiori) [inline src/main/kotlin/flag/FlagConfigUpdater.kt:26]: Nit: use javadoc comments for class/function comments, even for internal stuff.

```
/**
 *
 */
```

File: src/main/kotlin/flag/FlagConfigStreamApi.kt [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        reconnIntervalMillis)
+
+    internal fun connect() {
+        // Guarded by lock. Update to callbacks and waits can lead to race conditions.
+        lock.withLock {
+            val isInit = AtomicBoolean(true)
+            val connectTimeoutFuture = CompletableFuture<Unit>()
+            val updateTimeoutFuture = CompletableFuture<Unit>()
+            stream.onUpdate = { data ->
+                if (isInit.getAndSet(false)) {
```
Comment (bgiori) [inline src/main/kotlin/flag/FlagConfigStreamApi.kt:59]: This boolean logic feels like the opposite of what I would expect. Why would receiving an update set the `isInit` boolean to false? wouldn't an update mean that the stream is initialized?

File: src/main/kotlin/flag/FlagConfigStreamApi.kt [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                    // Resolve timeout.
+                    connectTimeoutFuture.complete(Unit)
+
+                    // Make sure valid data.
+                    try {
+                        val flags = getFlagsFromData(data)
+
+                        try {
+                            if (onInitUpdate != null) {
+                                onInitUpdate?.let { it(flags) }
```
Comment (bgiori) [inline src/main/kotlin/flag/FlagConfigStreamApi.kt:70]: use `?.invoke(flags)`

File: src/main/kotlin/flag/FlagConfigStreamApi.kt [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+                            } else {
+                                onUpdate?.let { it(flags) }
```
Comment (bgiori) [inline src/main/kotlin/flag/FlagConfigStreamApi.kt:72]: use `?.invoke(flags)`

File: src/main/kotlin/flag/FlagConfigStreamApi.kt [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        stream.cancel()
+    }
+
+    private fun getFlagsFromData(data: String): List<EvaluationFlag> {
+        return json.decodeFromString<List<EvaluationFlag>>(data)
+    }
+
+    private fun handleError(e: Exception?) {
+        close()
+        onError?.let { it(e) }
```
Comment (bgiori) [inline src/main/kotlin/flag/FlagConfigStreamApi.kt:144]: `?.invoke(e)`

File: src/main/kotlin/flag/FlagConfigStreamApi.kt [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    constructor(cause: Throwable?) : this(cause?.toString(), cause)
+}
+internal class FlagConfigStreamApiConnTimeoutError: FlagConfigStreamApiError("Initial connection timed out")
+internal class FlagConfigStreamApiDataCorruptError: FlagConfigStreamApiError("Stream data corrupted")
+internal class FlagConfigStreamApiStreamError(cause: Throwable?): FlagConfigStreamApiError("Stream error", cause)
+
+private const val CONNECTION_TIMEOUT_MILLIS_DEFAULT = 1500L
+private const val KEEP_ALIVE_TIMEOUT_MILLIS_DEFAULT = 17000L // keep alive sends at 15s interval. 2s grace period
+private const val RECONN_INTERVAL_MILLIS_DEFAULT = 15 * 60 * 1000L
+internal class FlagConfigStreamApi (
```
Comment (bgiori) [inline src/main/kotlin/flag/FlagConfigStreamApi.kt:31]: nit: newline before class def.

Comment (bgiori) [reply src/main/kotlin/flag/FlagConfigStreamApi.kt:31]: Also, we should all get in the habit of adding lightweight class comments for important classes even if they're internal.

File: src/main/kotlin/flag/FlagConfigUpdater.kt [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+    // Stop should stop updater temporarily. There may be another start in the future.
+    // To stop completely, with intention to never start again, use shutdown() instead.
+    fun stop()
+    // Destroy should stop the updater forever in preparation for server shutdown.
+    fun shutdown()
+}
+
+internal abstract class FlagConfigUpdaterBase(
```
Comment (bgiori) [inline src/main/kotlin/flag/FlagConfigUpdater.kt:34]: Add a quick class comment. What does this class do, what do extending classes modify and why.

File: src/main/kotlin/flag/FlagConfigUpdater.kt [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+        // Not guarded by lock. close() can cancel start().
+        flagConfigStreamApi.close()
+    }
+
+    override fun shutdown() = stop()
+}
+
+private const val RETRY_DELAY_MILLIS_DEFAULT = 15 * 1000L
+private const val MAX_JITTER_MILLIS_DEFAULT = 2000L
+internal class FlagConfigFallbackRetryWrapper(
```
Comment (bgiori) [inline src/main/kotlin/flag/FlagConfigUpdater.kt:193]: nit: newline before def

File: src/main/kotlin/util/SseStream.kt [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+import kotlin.math.max
+import kotlin.math.min
+
+internal class StreamException(error: String): Throwable(error)
+
+private const val KEEP_ALIVE_TIMEOUT_MILLIS_DEFAULT = 0L // no timeout
+private const val RECONN_INTERVAL_MILLIS_DEFAULT = 30 * 60 * 1000L
+private const val MAX_JITTER_MILLIS_DEFAULT = 5000L
+private const val KEEP_ALIVE_DATA = " "
+internal class SseStream (
```
Comment (bgiori) [inline src/main/kotlin/util/SseStream.kt:28]: nit: newline before def

File: src/main/kotlin/flag/FlagConfigUpdater.kt (hunk 11 above)
Comment (bgiori) [reply src/main/kotlin/flag/FlagConfigUpdater.kt:193]: nit: add class comment, what does this do?

File: src/main/kotlin/util/SseStream.kt [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+                return
+            }
+            cancel()
+            val err = t
+                ?: if (response != null) {
+                    StreamException(response.toString())
+                } else {
+                    StreamException("Unknown stream failure")
+                }
+            onError?.let { it(err) }
```
Comment (bgiori) [inline src/main/kotlin/util/SseStream.kt:97]: use invoke

**Review by bgiori — CHANGES_REQUESTED** [review]

File: src/main/kotlin/flag/FlagConfigStreamApi.kt (hunk 5 above)
Comment (zhukaihan) [reply src/main/kotlin/flag/FlagConfigStreamApi.kt]: It means if it's during an initialization process. Receiving an update terminates initialization. Renamed to `isDuringInit`.

**Review by bgiori — APPROVED** [review]

**Pull request merged**